src/Runner.Worker/ActionManager.cs

```diff
@@ -124,6 +132,81 @@ public async Task<PrepareResult> PrepareActionsAsync(IExecutionContext execution
                         }
                     }
 
+                    var repoAction = action.Reference as Pipelines.RepositoryPathReference;
+                    if (repoAction.RepositoryType != Pipelines.PipelineConstants.SelfAlias)
+                    {
+                        var definition = LoadAction(executionContext, action);
+                        if (definition.Data.Execution.HasPre)
+                        {
+                            var actionRunner = HostContext.CreateService<IActionRunner>();
+                            actionRunner.Action = action;
+                            actionRunner.Stage = ActionRunStage.Pre;
+                            actionRunner.Condition = definition.Data.Execution.InitCondition;
+
+                            Trace.Info($"Add 'pre' execution for {action.Id}");
+                            preStepTracker[action.Id] = actionRunner;
+                        }
+                    }
+                }
+                else if (action.Reference.Type == Pipelines.ActionSourceType.Repository && newActionMetadata)
+                {
+                    repositoryActions.Add(action);
+                }
+            }
+
+            if (repositoryActions.Count > 0)
+            {
+                // Get the download info
+                var downloadInfos = await GetDownloadInfoAsync(executionContext, repositoryActions);
+
+                // Download each action
+                foreach (var action in repositoryActions)
+                {
+                    var lookupKey = GetDownloadInfoLookupKey(action);
```

Comment:
Minor: `GetDownloadInfoAsyc` could just return a list of download info objects (i.e. the values of the dictionary it builds), so the for loop doesn't need to retrieve them by key. It should just loop over what was returned and download each one.

Reply:
Need to preserve the original list because of local actions (not downloaded). Oh wait local actions do need to be downloaded too.... crap let me think more about this.

Reply:
OK actually local actions are not downloaded. Ting reminded me, they come from checkout.